### PR TITLE
[Feature] support new compaction strategy for persistent index

### PR DESCRIPTION
**Pull request by luohaha**
More detail about this feature, check this #27581 

In this PR, it mainly includes the following modifications:

1. Introducing the l2 layer to transform merge compaction into minor compaction. By moving l1 to l2 each time, minor compaction can be accomplished. Another benefit of introducing l2 is to facilitate concurrent control between the apply thread and major compaction thread in the code.
2. Adding a major compaction thread pool to periodically merge multiple l2 files into one l2 file.

And also add three configs:
```
// Control if using the minor compaction strategy
enable_pindex_minor_compaction = true
// if l2 num is larger than this, stop doing minor compaction, add this config to prevent l2 grow too large.
max_allow_pindex_l2_num = 5
// control the background compaction threads, default value is equal to cpu number
pindex_major_compaction_num_threads = 0
// control the persistent index schedule compaction interval
pindex_major_compaction_schedule_interval_seconds = 15
```

By enable minor compaction, we can achieve better tail latency, thereby reducing overall latency fluctuations.

total record count | record count per upsert | disable minor compaction | enable major compaction 
-- | -- | -- | -- 
5000000 | 1000 | 2.485 s | 92 ms 
10000000 | 1000 | 5.069 s | 184 ms 
5000000 | 5000 | 2.452 s | 104 ms
10000000 | 5000 | 5.021 s | 192 ms

## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/storage/persistent_index.cpp [hunk 1]
```diff
@@ -3738,6 +3946,85 @@ Status PersistentIndex::_merge_compaction_internal(ImmutableIndexWriter* writer,
...
+        auto writer = std::make_unique<ImmutableIndexWriter>();
+        RETURN_IF_ERROR(writer->init(new_l1_filename, _version, true));
+        // followe this rules:
+        // 1, remove delete key when l2 not exist
+        // 2. skip merge l1, only merge tmp-l1 and l0
+        RETURN_IF_ERROR(_merge_compaction_internal(writer.get(), _has_l1 ? 1 : 0, _l1_vec.size(),
+                                                   _usage_and_size_by_key_length, _l2_vec.empty() ? false : true,
+                                                   nullptr));
+        RETURN_IF_ERROR(writer->finish());
+        LOG(INFO) << "PersistentIndex async compaction, merge tmp l1, merge cnt: " << _l1_vec.size()
```
Comment (chaoyli) [inline be/src/storage/persistent_index.cpp:3990]: Add the `tablet_id` in the log

File: be/src/storage/persistent_index.cpp [hunk 2]
```diff
@@ -3857,6 +4144,211 @@ Status PersistentIndex::_merge_compaction_advance() {
...
+    auto writer = std::make_unique<ImmutableIndexWriter>();
+    // use bigger edit version + 1 as new l2 edit version
+    EditVersion new_l2_version(l2_versions[l2_versions.size() - 1].major(),
+                               l2_versions[l2_versions.size() - 1].minor());
+    std::string print_str = "input: [";
+    for (const auto& l2_version : l2_versions) {
+        print_str += l2_version.to_string() + ",";
+    }
+    print_str += "], output: " + new_l2_version.to_string();
+    const std::string idx_file_path = strings::Substitute("$0/index.l2.$1.$2$3", _path, new_l2_version.major(),
```
Comment (chaoyli) [inline be/src/storage/persistent_index.cpp:4172]: Add the `tablet_id` in the log

File: be/src/storage/persistent_index.cpp [hunk 3]
```diff
@@ -2809,48 +2868,96 @@ Status PersistentIndex::abort() {
...
+    if (_has_l1) {
+        total_l1_l2_file_size += _l1_vec[0]->file_size();
+    }
+    for (int i = 0; i < _l2_vec.size(); i++) {
+        total_l1_l2_file_size += _l2_vec[i]->file_size();
+    }
+    return total_l1_l2_file_size;
+}
+
+bool PersistentIndex::_stop_do_async_compaction() {
```
Comment (chaoyli) [inline be/src/storage/persistent_index.cpp:2882]: Use the proactive name like `enable_async_compaction`

File: be/src/storage/persistent_index.cpp [hunk 4]
```diff
@@ -3738,6 +3946,85 @@ Status PersistentIndex::_merge_compaction_internal(ImmutableIndexWriter* writer,
...
+    return _has_l1 ? _l1_vec.size() - 1 : _l1_vec.size();
+}
+
+// There are a few steps in async compaction:
+// 1. flush l0 to l1:
+//    a. if there is only one tmp-l1 file, move this tmp-l1 to l1 file.
+//    b. if there are > 2 tmp-l1 file, then merge l0 and tmp-l1 files to new l1 file.
+//    c. if there is only one l1 file, flush l0 to new l1 file.
+//    d. if there is not l1 file exists, flush l0 to l1 file.
+// 2. move old l1 to l2. only if old l1 exist
```
Comment (chaoyli) [inline be/src/storage/persistent_index.cpp:3959]: Do you need to move the old l1 to l2, or just merge multiple l1 files to l2 when reach a limit?

Comment (luohaha) [reply be/src/storage/persistent_index.cpp:3959]: Move old l1 to l2 is better for apply latency.

File: be/src/storage/persistent_index.cpp (hunk 1 above)
Comment (luohaha) [reply be/src/storage/persistent_index.cpp:3990]: I think you can find tablet_id already in path ?

File: be/src/storage/persistent_index.cpp (hunk 3 above)
Comment (luohaha) [reply be/src/storage/persistent_index.cpp]: Done

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/storage/storage_engine.cpp [hunk 5]
```diff
@@ -898,6 +900,17 @@ Status StorageEngine::_perform_base_compaction(DataDir* data_dir, std::pair<int3
...
 
+Status StorageEngine::_perform_pk_index_bg_compaction(DataDir* data_dir) {
+    TabletSharedPtr best_tablet = _tablet_manager->find_best_tablet_to_do_pk_index_bg_compaction(data_dir);
+    if (best_tablet == nullptr) {
+        return Status::NotFound("there are no suitable tablets");
+    }
+    if (best_tablet->updates() == nullptr) {
+        return Status::InternalError("not an updatable tablet");
+    }
+    return best_tablet->updates()->pk_index_bg_compaction();
```
Comment (decster) [inline be/src/storage/storage_engine.cpp:911]: There may be multiple threads calling on the same dir, and pickup same tablet?
It seems you need a single thread to schedule compaction task and send picked up compaction tablet to threadpool?

File: be/src/storage/persistent_index.cpp [hunk 6]
```diff
@@ -3857,6 +4147,211 @@ Status PersistentIndex::_merge_compaction_advance() {
...
+            // empty l0 kvs
+            std::vector<KVRef> empty_l0_kvs;
+            RETURN_IF_ERROR(merge_shard_kvs(key_size, empty_l0_kvs, l2_kvs, estimate_size_per_shard, kvs));
+            // write shard
+            RETURN_IF_ERROR(writer->write_shard(key_size, npage_hint, nbucket, kvs));
+        }
+    }
+    RETURN_IF_ERROR(writer->finish());
+    _write_amp_score.store(0.0);
+    LOG(INFO) << "PersistentIndex compact l2 end : " << print_str << " cost: " << watch.elapsed_time();
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/persistent_index.cpp]: Done

File: be/src/storage/tablet_updates.cpp [hunk 7]
```diff
@@ -3719,6 +3724,33 @@ void TabletUpdates::get_basic_info_extra(TabletBasicInfo& info) {
     }
 }
 
+double TabletUpdates::get_pk_index_write_amp_score() {
+    double score = 0.0;
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/tablet_updates.cpp:3733]: Yes

File: be/src/storage/storage_engine.cpp (hunk 5 above)
Comment (luohaha) [reply be/src/storage/storage_engine.cpp]: Done, using threadpool to handle this.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

File: be/src/storage/tablet_manager.cpp [hunk 8]
```diff
@@ -687,6 +687,42 @@ TabletSharedPtr TabletManager::find_best_tablet_to_compaction(CompactionType com
...
+
+            pick_tablets.push_back(tablet_ptr);
+        }
+    }
+    // 2. sort tablet by score, by ascending order.
+    std::sort(pick_tablets.begin(), pick_tablets.end(), [](TabletSharedPtr& a, TabletSharedPtr& b) {
+        // We try to compact tablet with small write amplification score first,
+        // to improve the total write IO amplification
+        return a->updates()->get_pk_index_write_amp_score() < b->updates()->get_pk_index_write_amp_score();
+    });
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/tablet_manager.cpp]: Yes, fixed.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by decster — DISMISSED** [review]

File: be/src/storage/persistent_index.cpp [hunk 9]
```diff
@@ -3738,6 +3949,85 @@ Status PersistentIndex::_merge_compaction_internal(ImmutableIndexWriter* writer,
...
+    }
+    // 2. move old l1 to l2.
+    if (_has_l1) {
+        // just link old l1 file to l2
+        const std::string l2_file_path =
+                strings::Substitute("$0/index.l2.$1.$2", _path, _l1_version.major(), _l1_version.minor());
+        const std::string old_l1_file_path =
+                strings::Substitute("$0/index.l1.$1.$2", _path, _l1_version.major(), _l1_version.minor());
+        LOG(INFO) << "PersistentIndex async compaction, link from " << old_l1_file_path << " to " << l2_file_path;
+        RETURN_IF_ERROR(FileSystem::Default()->link_file(old_l1_file_path, l2_file_path));
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/persistent_index.cpp:4016]: Not need that, old_l1_file_path will be remove later.

File: be/src/storage/persistent_index.cpp (hunk 1 above)
Comment (luohaha) [reply be/src/storage/persistent_index.cpp]: Yes

**Review by chaoyli — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by decster — APPROVED** [review]

**Pull request merged**